libsql-sqlite3/src/vectordiskann.c

```diff
@@ -0,0 +1,345 @@
+  return (u64)p[0]
+       | (u64)p[1] << 8
+       | (u64)p[2] << 16
+       | (u64)p[3] << 24
+       | (u64)p[4] << 32
+       | (u64)p[5] << 40
+       | (u64)p[6] << 48
+       | (u64)p[7] << 56;
+}
+
+static inline void writeLE16(unsigned char *p, u16 v){
+  p[0] = v;
+  p[1] = v >> 8;
+}
+
+static inline void writeLE64(unsigned char *p, u64 v){
+  p[0] = v;
+  p[1] = v >> 8;
+  p[2] = v >> 16;
+  p[3] = v >> 24;
+  p[4] = v >> 32;
+  p[5] = v >> 40;
+  p[6] = v >> 48;
+  p[7] = v >> 56;
+}
+
+/**************************************************************************
+** BlobSpot utilities
+**************************************************************************/
+
+// sqlite3_blob_* API return SQLITE_ERROR in any case but we need to distinguish between "row not found" and other errors in some cases
+static int blobSpotConvertRc(const DiskAnnIndex *pIndex, int rc){
+  if( rc == SQLITE_ERROR && strncmp(sqlite3_errmsg(pIndex->db), "no such rowid", 13) == 0 ){
+    return DISKANN_ROW_NOT_FOUND;
+  }
+  return rc;
+}
+
+int blobSpotCreate(const DiskAnnIndex *pIndex, BlobSpot **ppBlobSpot, u64 nRowid, int nBufferSize, int isWritable) {
+  int rc = SQLITE_OK;
+  BlobSpot *pBlobSpot;
+  u8 *pBuffer;
+
+  DiskAnnTrace(("blob spot created: rowid=%lld, isWritable=%d\n", nRowid, isWritable));
+  assert( nBufferSize > 0 );
+
+  pBlobSpot = sqlite3_malloc(sizeof(BlobSpot));
+  if( pBlobSpot == NULL ){
+    rc = SQLITE_NOMEM_BKPT;
+    goto out;
+  }
+
+  pBuffer = sqlite3_malloc(nBufferSize);
+  if( pBuffer == NULL ){
+    rc = SQLITE_NOMEM_BKPT;
+    goto out;
+  }
+
+  // open blob in the end so we don't need to close it in error case
+  rc = sqlite3_blob_open(pIndex->db, pIndex->zDb, pIndex->zShadow, "data", nRowid, isWritable, &pBlobSpot->pBlob);
+  rc = blobSpotConvertRc(pIndex, rc);
+  if( rc != SQLITE_OK ){
+    goto out;
+  }
+  pBlobSpot->nRowid = nRowid;
+  pBlobSpot->pBuffer = pBuffer;
+  pBlobSpot->nBufferSize = nBufferSize;
+  pBlobSpot->isWritable = isWritable;
+  pBlobSpot->isInitialized = 0;
+
+  *ppBlobSpot = pBlobSpot;
+  return SQLITE_OK;
+
+out:
+  if( pBlobSpot != NULL ){
+    sqlite3_free(pBlobSpot);
+  }
+  if( pBuffer != NULL ){
+    sqlite3_free(pBuffer);
+  }
+  return rc;
+}
+
+int blobSpotReload(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, u64 nRowid, int nBufferSize) {
+  int rc;
+
+  DiskAnnTrace(("blob spot reload: rowid=%lld\n", nRowid));
+  assert( pBlobSpot != NULL && pBlobSpot->pBlob != NULL );
+  assert( pBlobSpot->nBufferSize == nBufferSize );
+
+  if( pBlobSpot->nRowid == nRowid && pBlobSpot->isInitialized ){
+    return SQLITE_OK;
+  }
+
+  if( pBlobSpot->nRowid != nRowid ){
+    rc = sqlite3_blob_reopen(pBlobSpot->pBlob, nRowid);
+    rc = blobSpotConvertRc(pIndex, rc);
+    if( rc != SQLITE_OK ){
+      return rc;
+    }
+  }
+  rc = sqlite3_blob_read(pBlobSpot->pBlob, pBlobSpot->pBuffer, nBufferSize, 0);
+  if( rc != SQLITE_OK ){
+    return rc;
+  }
+  pBlobSpot->isInitialized = 1;
```

Comment:
Should we update `pBlobSpot->nRowid` to `nRowid` here?

Reply:
or maybe better do:
```
  if( pBlobSpot->nRowid != nRowid ){
    rc = sqlite3_blob_reopen(pBlobSpot->pBlob, nRowid);
    rc = blobSpotConvertRc(pIndex, rc);
    if( rc != SQLITE_OK ){
      return rc;
    }
    pBlobSpot->nRowid = nRowid;
    pBlobSpot->isInitialized = 0;
  }
```
in lines 173-179

Reply:
Sure, this is bug :bug: !